File: searchRange.cpp
// Time Complexity : O(logn) 
// Space Complexity : O(1)  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// [***] KEY INSIGHT: FIND FIRST INDEX USING BS AND THEN FIND LAST INDEX

// 1. Finding first index: Find the target using BS and once found see if it is same as left element; if no its the first index
// 2. If it is same as left then make high=mid-1
// 3. Repeat similar process to find last index 

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> result = {-1,-1};
        if(nums.size() == 0)
            return result;
        int low = 0, high = nums.size()-1, first=-1, last=-1;
        while(low<=high){
            int mid = low+(high-low)/2;
            if(nums[mid] == target){
                if(mid-1>=low && nums[mid-1] == target)
                    high = mid-1;
                else{
                    first = mid;
                    break;
                }
            }
            else{
                if(target<nums[mid])
                    high=mid-1;
                else
                    low = mid+1;
            }
        }
        low = 0; high = nums.size()-1;
        while(low<=high){
            int mid = low+(high-low)/2;
            if(nums[mid] == target){
                if(mid+1<=high && nums[mid+1] == target)
                    low = mid+1;
                else{
                    last = mid;
                    break;
                }
            }
            else{
                if(target<nums[mid])
                    high=mid-1;
                else
                    low = mid+1;
            }
        }
        
        result[0] = first;
        result[1] = last;
        return result;
    }
};

File: SearchRange.py

# time complexity : O(log(n))
# space complexity : O(1)
class Solution:
    def searchRange(self, nums, target):
       
        n = len(nums)
        def bs(nums, element, isFirst):
            index = -1
            low, high = 0, n-1
            while low <= high:
                mid = (low+high)//2
                num = nums[mid]
                if num == element and isFirst:
                    high = mid -1
                    index = mid
                elif num == element and not isFirst:
                    low = mid + 1
                    index = mid
                elif num > target:
                    high = mid -1
                else:
                    low = mid + 1
            return index

        return [bs(nums, target, True), bs(nums, target, False)]